cypress-io/cypress
# feat: ACI Notification Banners

- Closes 
#22933 
#22934 
#22798 
#22935 

### User facing changelog
* Add new banners in the app to help get your project connected to the dashboard and complete a first run recording

### Additional details
Four main areas here:
1. Adding new section of project preferences to track the "last shown" and "dismissed" timestamps for each banner being added here
2. Create new `TrackedBanner` component to allow reuse of storing "last shown" and "dismissed" info for banners
3. Add four new "Smart Notifications" that trigger based on users progress towards connecting their project to the dashboard
	* Login
	    * Has been using Cypress for at least 4 days
	    * Is **_not_** logged in
	    * Not previously dismissed in this project
	* Create Organization
	    * Has been using Cypress for at least 4 days
	    * Is logged in
	    * Is **_not_** a member of any cloud organizations
	    * Not previously dismissed in this project
	* Connect Project
	    * Has been using Cypress for at least 4 days
	    * Is logged in
	    * Is member of at least one cloud organization
	    * Project is **_not_** successfully connected to dashboard
	    * Not previously dismissed in this project 
	* Record First Run 	
	    * Has been using Cypress for at least 4 days
	    * Is logged in
	    * Is member of at least one cloud organization
	    * Project is successfully connected to dashboard
	    * Project does **_not_** have any cloud recording runs
	    * Not previously dismissed in this project
4. Connect new banners to SpecsList in the `SpecsListBanners` component - this is currently serving as a "banner controller" of sorts to control which banners to display and when (a refinement of this construct is coming in future work related to A/B testing of banner content)

### Steps to test
**Note: To fully test it will be necessary to either create a new user or have an existing user removed from all organizations. You may want to create a new "throwaway" user rather than remove yourself from the shared Cypress org in staging.**

**Note: launch with CYPRESS_INTERNAL_ENV=staging for staging**.

1. Create a new empty Cypress project (name it something distinctive) with at least one spec (content is unimportant)
2. Open the project in Cypress. Ensure you are logged out. Validate that "Smart Login Banner" does not display.
3. Close the project and find its `state.json` file. On Mac this is stored in a project-specific directory under `~/Library/Application Support/Cypress/cy/staging/projects/`
4. Modify `state.json` to have a `firstOpened` value more than 4 days in the past: `Date.now() - 8.64e+7 * 4`
5. Reopen Cypress, validate the "Smart Login Banner" now displays
6. Dismiss the banner by clicking the "X" button. Refresh app and validate banner does not display
7. Reopen `state.json`, observe an entry for the login banner has been created. Remove it, and reopen the project.
```
    "aci_082022_login": {
      "lastShown": 1660075007022,
      "dismissed": 1660075012749
    },
```
9. Observe login banner once again displays. Use banner to execute login flow, validate it succeeds.
   * Note: Next three prompts use same logic for respecting dismissal
11. "Smart Create Organization Banner" should now display (if your user has been removed from all cloud orgs). Use banner to create an organization, validate UI updates to no longer show create organization banner
12. "Smart Connect Project Banner" should now display. Use banner to create new cloud project, validate UI updates to no longer show connect project banner
13. "Smart Record Banner" should now display. Validate record command looks correct (should include `--component` flag for CT, valid record key) and can be copy-pasted
14. _Optional:_ Record a cloud run, validate that "Smart Record Banner" does not display on next app launch

### How has the user experience changed?
**Login Banner**

https://user-images.githubusercontent.com/11482842/183981545-0f78a87c-6ad9-4c9f-a8a0-4e2becd2a1be.mov

**Create Org Banner**

https://user-images.githubusercontent.com/11482842/183981573-78b599bd-f2a9-48b8-b686-1c57930742e7.mov

**Connect Project Banner**

https://user-images.githubusercontent.com/11482842/183981601-0dade474-003a-4838-afdf-77aae5fca8ed.mov

**Record Banner**

https://user-images.githubusercontent.com/11482842/183981630-8233f4a1-a229-4890-8405-591b30d8d895.mov

### PR Tasks

- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
